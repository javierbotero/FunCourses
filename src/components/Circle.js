import React from 'react';
import PropTypes from 'prop-types';

const Circle = props => {
  const {
    quantity,
  } = props;

  const width = 80;
  const height = 80;
  const strWidth = 6;
  const radius = (width / 2) - (strWidth * 2);
  const circumference = radius * 2 * Math.PI;
  const strCircumference = `${circumference} ${circumference}`;
  const completed = circumference - ((circumference / 100) * quantity);

  return (
    <div className="book-child records">
      <svg style={{
        width,
        height,
      }}
      >
        <defs>
          <radialGradient id="gradient" cx="50%" cy="50%" r="70%">
            <stop offset="60%" stopColor="#4333B3" />
            <stop offset="110%" stopColor="#E04534" />
          </radialGradient>
        </defs>
        <circle
          r={radius}
          cx={width / 2}
          cy={height / 2}
          stroke="#e8e8e8"
          style={{
            strokeWidth: strWidth,
            fill: 'transparent',
          }}
        />
        <circle
          r={radius}
          cx={width / 2}
          cy={height / 2}
          stroke="url(#gradient)"
          style={{
            strokeWidth: strWidth,
            strokeDasharray: strCircumference,
            strokeDashoffset: completed,
            fill: 'transparent',
            transform: 'rotate(-90deg)',
            transformOrigin: 'center',
          }}
        />
      </svg>
    </div>
  );
};

Circle.propTypes = {
  quantity: PropTypes.number.isRequired,
};

export default Circle;
