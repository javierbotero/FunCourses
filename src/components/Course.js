import React from 'react';
import { Link } from 'react-router-dom';
import ReactRouterPropTypes from 'react-router-prop-types';
import PropTypes from 'prop-types';

const Course = props => {
  const {
    match,
    courses,
    location,
    url,
  } = props;
  const course = courses.find(c => c.id === parseInt(match.params.id, 10));

  return (
    <div>
      <nav>
        <ul>
          <li>
            <Link to={location.state ? location.state.from.pathname : `${url}`}>
              &#60;
            </Link>
          </li>
        </ul>
      </nav>
      <div>
        <div>{course.title}</div>
        <div>
          <Link
            to={{
              pathname: `${url}/user/${course.teacher_id}`,
              state: { from: location },
            }}
          >
            Teacher:
            {` ${course.teacher.username}`}
          </Link>
        </div>
      </div>
      <div>{course.content}</div>
      <div>
        status:
        {` ${course.status}`}
      </div>
    </div>
  );
};

Course.propTypes = {
  match: ReactRouterPropTypes.match.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
  courses: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    teacher_id: PropTypes.number.isRequired,
    status: PropTypes.string.isRequired,
    dates: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
    teacher: PropTypes.shape({
      id: PropTypes.number,
      username: PropTypes.string,
    }).isRequired,
    favorites: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number,
    })).isRequired,
    subscriptions: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number.isRequired,
      confirmed: PropTypes.bool,
    })).isRequired,
    comments: PropTypes.arrayOf(PropTypes.shape({
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      body: PropTypes.string,
    })).isRequired,
    pendings: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      confirmed: PropTypes.bool,
    })).isRequired,
  })).isRequired,
  url: PropTypes.string.isRequired,
};

export default Course;
