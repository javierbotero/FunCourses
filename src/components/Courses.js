import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import CoursesCss from '../css/Courses.module.css';
import avatarImage from '../assets/images/avatar.jpg';

const Courses = props => {
  const {
    courses,
    coursesToDivs,
    location,
    url,
    useAuth,
    resetUser,
    resetCourses,
    avatar,
    username,
    urlApi,
    setShowMenu,
  } = props;
  const popUpMenu = () => {
    const div = document.getElementsByClassName('scrollContainer')[0];
    const navMenu = () => document.querySelector('.navMenu');
    if (div.scrollLeft < 200) {
      navMenu().classList.add('hide');
      navMenu().classList.remove('show');
    } else {
      navMenu().classList.remove('hide');
      navMenu().classList.add('show');
    }
  };
  useEffect(() => {
    const div = document.getElementsByClassName('scrollContainer')[0];
    div.addEventListener('scroll', popUpMenu);
    return () => {
      div.removeEventListener('scroll', popUpMenu);
    };
  }, []);
  const authObject = useAuth();
  const logout = () => {
    localStorage.removeItem('currentUserIdFunCourses');
    localStorage.removeItem('currentUserPasswordFunCourses');
    authObject.setUserId(null);
    authObject.setUserPassword(null);
    resetUser();
    resetCourses();
  };

  return (
    <div
      className={`${CoursesCss.courses} scrollContainer`}
    >
      <nav className="navMenu hide">
        <button className="removeButtonStyles" type="button" onClick={() => setShowMenu(true)}>
          <FontAwesomeIcon icon="bars" />
        </button>
      </nav>
      <header className={`${CoursesCss.coursesMenu}`}>
        <div className={`${CoursesCss.avatarZone}`}>
          <div
            className={`avatar ${CoursesCss.avatarPic}`}
            style={{
              backgroundImage: avatar ? `url(${urlApi}${avatar.slice(1, avatar.length)})` : `url(${avatarImage})`,
            }}
          />
          <h4 className={CoursesCss.username}>{username}</h4>
        </div>
        <ul className={`${CoursesCss.menu}`}>
          <li>
            <Link to={{
              pathname: `${url}/dashboard`,
              state: { from: location },
            }}
            >
              Dashboard
            </Link>
          </li>
          <li>
            <Link to={{
              pathname: `${url}/create`,
              state: { from: location },
            }}
            >
              Create
            </Link>
          </li>
          <li>
            <Link to={{
              pathname: `${url}/report`,
              state: { from: location },
            }}
            >
              Report
            </Link>
          </li>
        </ul>
        <button className={`${CoursesCss.btn}`} type="button" onClick={logout}>Log out</button>
      </header>
      <div className={`${CoursesCss.coursesItems}`}>
        {coursesToDivs(courses, url, location)}
      </div>
    </div>
  );
};

Courses.propTypes = {
  courses: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    teacher_id: PropTypes.number.isRequired,
    status: PropTypes.string.isRequired,
    dates: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
    teacher: PropTypes.shape({
      id: PropTypes.number,
      username: PropTypes.string,
    }).isRequired,
    favorites: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number,
    })).isRequired,
    subscriptions: PropTypes.arrayOf(PropTypes.shape({
      course_id: PropTypes.number,
      user_id: PropTypes.number.isRequired,
      confirmed: PropTypes.bool,
    })).isRequired,
    comments: PropTypes.arrayOf(PropTypes.shape({
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      body: PropTypes.string,
    })),
    pendings: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number,
      user_id: PropTypes.number,
      course_id: PropTypes.number,
      confirmed: PropTypes.bool,
    })).isRequired,
  })).isRequired,
  coursesToDivs: PropTypes.func.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
  url: PropTypes.string.isRequired,
  useAuth: PropTypes.func.isRequired,
  resetUser: PropTypes.func.isRequired,
  resetCourses: PropTypes.func.isRequired,
  avatar: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  urlApi: PropTypes.string.isRequired,
  setShowMenu: PropTypes.func.isRequired,
};

export default Courses;
