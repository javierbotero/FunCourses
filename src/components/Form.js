import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ReactRouterPropTypes from 'react-router-prop-types';
import {
  setUserError,
  removeUserError,
  setTrueLoading,
  setFalseLoading,
} from '../reducers/user';
import formCss from '../css/Form.module.css';
import fun1 from '../assets/images/fun1.jpeg';
import fun2 from '../assets/images/fun2.jpg';

const Form = props => {
  const {
    initCreator,
    tokenPayload,
    id,
    token,
    url,
    handleApiRequest,
    match,
    setUserErr,
    removeUserErr,
    setTrueLoad,
    setFalseLoad,
    history,
    authenticate,
  } = props;
  const [userName, setUserName] = useState('');
  const [password, setPassword] = useState('');
  const [passConf, setPassConf] = useState('');
  const [email, setEmail] = useState('');
  const [avatar, setAvatar] = useState('');
  const [matcher, setMatcher] = useState(match.params.identifier);
  const handleSubmit = async e => {
    e.preventDefault();
    setTrueLoad();
    if (matcher === 'signup') {
      if (password.length < 5 || userName.length < 5 || email.length < 5 || password !== passConf) {
        setUserErr('Please fill the form correctly, Username and password should be more than 4 characters, password and confirmation must be equal');
        return;
      }
    } else if (password.length < 5 || userName.length < 5) {
      setUserErr('Please fill the form correctly, Username and password should be more than 4 characters');
      return;
    }
    const tokenInfo = tokenPayload(id, token);
    const dataSignUp = {
      ...tokenInfo,
      avatar,
      user: {
        username: userName,
        password,
        password_confirmation: passConf,
        email,
      },
    };
    const dataLogIn = {
      ...tokenInfo,
      username: userName,
      password,
    };
    authenticate(
      handleApiRequest,
      initCreator,
      'POST',
      `${url}${matcher === 'signup' ? 'signup' : 'login'}`,
      matcher === 'signup' ? dataSignUp : dataLogIn,
      setFalseLoad,
    );
    setUserName('');
    setEmail('');
    setPassword('');
    setPassConf('');
  };
  const handleChange = (e, cb, attach = false) => {
    removeUserErr();
    if (attach && e.target.files[0]) {
      const pictureData = new FileReader();
      pictureData.onloadstart = () => setTrueLoad();
      pictureData.onloadend = () => setFalseLoad();
      pictureData.readAsDataURL(e.target.files[0]);
      pictureData.onload = () => {
        cb(
          {
            io: pictureData.result.split(',')[1],
            filename: e.target.files[0].name,
          },
        );
      };
      pictureData.onerror = () => {
        setUserErr('File could not be loaded');
      };
    } else {
      cb(e.target.value);
    }
  };
  const changeHistory = str => {
    setMatcher(str);
    history.push(`/${str}`);
  };

  const infoSign = (signup = false) => {
    if (signup) {
      return (
        <div>
          {'Hello there! join our community and start '}
          {'enjoying fun courses. If you have an account please '}
          <button className="button-3" type="button" onClick={() => changeHistory('login')}>Log in</button>
        </div>
      );
    }
    return (
      <div>
        {'Hello there! it\'s time to learn with fun courses. '}
        {"If you don't have an account please "}
        <button className="button-3" type="button" onClick={() => changeHistory('signup')}>Sign up</button>
      </div>
    );
  };

  return (
    <div
      className="form positionImage"
      style={{ backgroundImage: matcher === 'signup' ? `url(${fun1})` : `url(${fun2})` }}
    >
      <div className={`${formCss.layer} ${matcher === 'signup' ? formCss.signup : formCss.login}`} />
      <h3 className={`${formCss.index2} form-item`}>{matcher === 'signup' ? 'Sign Up' : 'Log in'}</h3>
      <div className={`form-item ${formCss.index2}`}>{matcher === 'signup' ? infoSign(true) : infoSign()}</div>
      <form className={`form-item ${formCss.index2}`} onSubmit={handleSubmit}>
        <div>
          <label htmlFor="username">
            <div>Username</div>
            <input type="text" onChange={e => handleChange(e, setUserName)} value={userName} id="username" placeholder="User" />
          </label>
        </div>
        {matcher === 'signup'
          && (
            <div>
              <label htmlFor="email">
                <div>Email</div>
                <input type="email" onChange={e => handleChange(e, setEmail)} value={email} id="email" placeholder="user@mail.com" />
              </label>
            </div>
          )}
        <div>
          <label htmlFor="password">
            <div>Password</div>
            <input type="password" onChange={e => handleChange(e, setPassword)} value={password} id="password" placeholder="password" />
          </label>
        </div>
        { matcher === 'signup' && (
          <div>
            <label htmlFor="passwordConfirmation">
              <div>Password confirmation</div>
              <input type="password" onChange={e => handleChange(e, setPassConf)} value={passConf} id="passwordConfirmation" placeholder="confirmation" />
            </label>
          </div>
        )}
        { matcher === 'signup' && (
          <div>
            <label htmlFor="avatar">
              <div>Image Profile</div>
              <input type="file" alt="avatar" onChange={e => handleChange(e, setAvatar, true)} id="avatar" name="avatar" accept="image/*" />
            </label>
          </div>
        )}
        <div className={`${formCss.buttonSubmit}`}>
          <input type="submit" value="Submit" className="btn button-2" />
        </div>
      </form>
    </div>
  );
};

Form.propTypes = {
  initCreator: PropTypes.func.isRequired,
  tokenPayload: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  handleApiRequest: PropTypes.func.isRequired,
  match: ReactRouterPropTypes.match.isRequired,
  setUserErr: PropTypes.func.isRequired,
  removeUserErr: PropTypes.func.isRequired,
  setTrueLoad: PropTypes.func.isRequired,
  setFalseLoad: PropTypes.func.isRequired,
  history: ReactRouterPropTypes.history.isRequired,
  authenticate: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  setUserErr: err => dispatch(setUserError(err)),
  removeUserErr: () => dispatch(removeUserError()),
  setTrueLoad: () => dispatch(setTrueLoading()),
  setFalseLoad: () => dispatch(setFalseLoading()),
});

export default connect(null, mapDispatchToProps)(Form);
export { Form };
