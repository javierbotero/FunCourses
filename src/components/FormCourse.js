import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { createCourse } from '../actions/interactions';
import {
  setUserError,
  setTrueLoading,
  setFalseLoading,
} from '../reducers/user';
import image from '../assets/images/fun1.jpeg';
import form from '../css/Form.module.css';
import styles from '../css/FormCourse.module.css';

const FormCourse = props => {
  const {
    userId,
    userPassword,
    username,
    location,
    urlApi,
    objThunk,
    userPayload,
    tokenPayload,
    id,
    token,
    dispatchCourse,
    setUserErr,
    setTrueLoad,
    setFalseLoad,
    url,
    setShowMenu,
  } = props;
  useEffect(() => {
    const deskTopMenu = document.querySelector('.desktopMenu');
    if (deskTopMenu.classList.contains('hide')) {
      deskTopMenu.classList.remove('hide');
    }
  });
  const [title, setTitle] = useState('');
  const [content, setContent] = useState('');
  const [dateStart, setDateStart] = useState(new Date().toLocaleDateString());
  const [dateEnd, setDateEnd] = useState(new Date().toLocaleDateString());
  const [price, setPrice] = useState('');
  const [provider, setProvider] = useState('zoom');
  const [link, setLink] = useState('');
  const [status, setStatus] = useState('Closed');
  const [mainPicture, setMainPicture] = useState('');
  const [pictures, setPictures] = useState([]);

  const handleSubmit = e => {
    e.preventDefault();
    const init = {
      ...userPayload(userId, userPassword),
      ...tokenPayload(id, token),
      main: mainPicture,
      images: pictures,
      course: {
        link,
        provider,
        title,
        content,
        status,
        dates: `${dateStart} ${dateEnd} `,
        price,
      },
    };
    const payload = objThunk(urlApi, 'POST', init);
    dispatchCourse(payload);
    setTitle('');
    setContent('');
    setDateStart(new Date().toLocaleDateString());
    setDateEnd(new Date().toLocaleDateString());
    setPrice('');
    setProvider('zoom');
    setLink('');
    setStatus('Closed');
    setMainPicture('');
    setPictures([]);
  };
  const saveFileToBase64 = (file, cb, cb2 = false) => {
    const newFile = new FileReader();
    newFile.onloadstart = () => { setTrueLoad(); };
    newFile.onloadend = () => { setFalseLoad(); };
    newFile.readAsDataURL(file);
    newFile.onload = () => {
      if (cb2) {
        cb2({
          io: newFile.result.split(',')[1],
          filename: file.name,
        }, cb);
      } else {
        cb({
          io: newFile.result.split(',')[1],
          filename: file.name,
        });
      }
    };
  };
  const handleChange = (e, cb, attach = false) => {
    if (attach) {
      if (e.target.files.length > 3) {
        setUserErr('Maximum number of pictures is 3');
      } else if (e.target.files.length > 1) {
        const pushToStrs = (obj, locCb) => locCb(state => [...state, obj]);
        Array.prototype
          .forEach
          .call(e.target.files, f => saveFileToBase64(f, cb, pushToStrs));
      } else if (e.target.files[0]) {
        saveFileToBase64(e.target.files[0], cb);
      }
    } else {
      cb(e.target.value);
    }
  };
  return (
    <div
      className={`${styles.container} positionImage`}
      style={{
        backgroundImage: `url(${image})`,
      }}
    >
      <nav className={`${styles.navBar} navMenu show`}>
        <div>
          <Link to={{
            pathname: location.state ? location.state.from.pathname : `${url}`,
            state: { from: location },
          }}
          >
            <FontAwesomeIcon icon="chevron-left" />
          </Link>
        </div>
        <div>
          Create
        </div>
        <button
          className="removeButtonStyles"
          type="button"
          onClick={() => {
            console.log('fired button menu dots');
            setShowMenu(true);
          }}
        >
          <FontAwesomeIcon icon="ellipsis-v" />
        </button>
      </nav>
      <div className={`${form.layer} ${styles.bgImage}`} id="layer" />
      <div className={styles.form}>
        <header className={`${styles.header}`}>
          <div>
            <h3>
              {username}
            </h3>
            <h4>create a fun course!</h4>
          </div>
        </header>
        <form className={`form ${styles.correctForm}`} onSubmit={handleSubmit}>
          <div>
            <label htmlFor="title">
              <div>Title</div>
              <input type="text" onChange={e => handleChange(e, setTitle)} value={title} id="title" placeholder="Title" />
            </label>
          </div>
          <div>
            <label htmlFor="content">
              <div>Content</div>
              <textarea
                type="text"
                onChange={e => handleChange(e, setContent)}
                value={content}
                id="content"
                placeholder="Be concise"
                className="long-text"
              />
            </label>
          </div>
          <div>
            <label htmlFor="dateStart">
              <div>Start</div>
              <input type="date" onChange={e => handleChange(e, setDateStart)} value={dateStart} id="dateStart" max={dateEnd} />
            </label>
            <label htmlFor="dateEnd">
              <div>End</div>
              <input type="date" onChange={e => handleChange(e, setDateEnd)} value={dateEnd} id="dateEnd" min={dateStart} />
            </label>
          </div>
          <div>
            <label htmlFor="price">
              <div>Price</div>
              <input type="text" onChange={e => handleChange(e, setPrice)} value={price} id="price" placeholder="Price" />
            </label>
          </div>
          <div>
            <label htmlFor="price">
              <div>Videocall provider</div>
              <select value={provider} onChange={e => handleChange(e, setProvider)}>
                <option value="zoom">Zoom</option>
                <option value="meet">Meet</option>
                <option value="livestorm">Livestorm</option>
                <option value="zoho-meeting">Zoho Meeting</option>
                <option value="pexip">Pexip</option>
                <option value="poly">Poly</option>
                <option value="Blue-jeans">Blue Jeans</option>
              </select>
            </label>
          </div>
          <div>
            <label htmlFor="link">
              <div>link</div>
              <input type="text" onChange={e => handleChange(e, setLink)} value={link} id="link" placeholder="Paste the link" />
            </label>
          </div>
          <div>
            <label htmlFor="link">
              <div>Status</div>
              <select value={status} onChange={e => handleChange(e, setStatus)}>
                <option value="Open">Open</option>
                <option value="Closed">Closed</option>
              </select>
            </label>
          </div>
          <div>
            <label htmlFor="main-picture">
              <div>Main picture</div>
              <input type="file" id="main-picture" name="main-picture" onChange={e => handleChange(e, setMainPicture, true)} accept="image/*" />
            </label>
          </div>
          <div>
            <label htmlFor="main-picture">
              <div>Add 3 more engaging pictures</div>
              <input type="file" id="main-picture" name="main-picture" onChange={e => handleChange(e, setPictures, true)} multiple accept="image/*" />
            </label>
          </div>
          <div className={styles.submit}>
            <input type="submit" value="Submit" className="button" />
          </div>
        </form>
      </div>
    </div>
  );
};

FormCourse.propTypes = {
  userId: PropTypes.number.isRequired,
  userPassword: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
  dispatchCourse: PropTypes.func.isRequired,
  urlApi: PropTypes.string.isRequired,
  objThunk: PropTypes.func.isRequired,
  userPayload: PropTypes.func.isRequired,
  tokenPayload: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  token: PropTypes.string.isRequired,
  setUserErr: PropTypes.func.isRequired,
  setTrueLoad: PropTypes.func.isRequired,
  setFalseLoad: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
  setShowMenu: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  dispatchCourse: payload => dispatch(createCourse(payload)),
  setUserErr: payload => dispatch(setUserError(payload)),
  setTrueLoad: () => dispatch(setTrueLoading()),
  setFalseLoad: () => dispatch(setFalseLoading()),
});

export default connect(null, mapDispatchToProps)(FormCourse);
export { FormCourse };
