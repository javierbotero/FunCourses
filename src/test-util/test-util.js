import fetchMock from 'fetch-mock';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { URL } from '../constants/constants';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mockMatch = {
  params: { id: '5' },
  path: 'some-path',
  url: 'url',
};
const mockCourses = [
  {
    id: 5,
    title: 'Karate class',
    content: 'Nice Karate class',
    teacher_id: 7,
    status: 'Open',
    dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
    price: 10,
    images_url: [],
    created_at: '2021-03-07T17:56:22.487Z',
    updated_at: '2021-03-15T04:21:24.489Z',
    teacher: {
      id: 7,
      username: 'Adele',
    },
    pendings: [],
    pending_students: [],
    favorites: [],
    subscriptions: [],
    confirmed_students: [],
  },
];
const mockHistory = {
  action: 'PUSH',
  block: jest.fn(),
  createHref: jest.fn(),
  go: jest.fn(),
  listen: jest.fn(),
  push: jest.fn(),
  location: {
    hash: '',
    pathname: '',
    search: '',
  },
  replace: jest.fn(),
};
const mockUser = {
  id: 7,
  status: 'Student',
  username: 'Javier 4',
  url_avatar: 'api.com/users/1',
  courses_as_student: [],
  pending_courses_as_student: [],
  comments: [],
  favorites: [],
  pending_to_accept_friendships: [],
  pending_requested_friendships: [],
  friendship_requests: [],
  requests: [],
  pendings: [],
  courses: [
    {
      id: 5,
      title: 'Karate class',
      content: 'Nice Karate class',
      teacher_id: 7,
      status: 'Open',
      dates: '2021-03-25T00:00:00.000Z 2021-03-29T00:00:00.000Z ',
      price: 10,
      created_at: '2021-03-07T17:56:22.487Z',
      updated_at: '2021-03-15T04:21:24.489Z',
      teacher: {
        id: 7,
        username: 'Adele',
      },
      pendings: [],
      pending_students: [],
      favorites: [],
      subscriptions: [],
      confirmed_students: [],
    },
  ],
};
const mockTeacherCourses = [];
const mockLocation = {
  from: { pathname: 'url' },
  hash: 'some-hash',
  pathname: 'url',
  search: 'default',
  state: {
    user: {
      id: 1,
      username: 'Paco',
    },
    from: {
      pathname: 'url',
    },
  },
};
const mockUrl = 'url';
const mockUsersListToDiv = jest.fn();
const mockCommentsToDivs = jest.fn();
const userPassword = '12345';
const func = jest.fn();
const mockUrlApi = 'url';
const mockToken = 'token';
const mockId = 1;

export {
  func,
  mockStore,
  fetchMock,
  mockMatch,
  mockCourses,
  mockTeacherCourses,
  mockLocation,
  mockUrl,
  mockUsersListToDiv,
  mockCommentsToDivs,
  mockUrlApi,
  mockToken,
  mockId,
  mockUser,
  URL,
  mockHistory,
  userPassword,
};
