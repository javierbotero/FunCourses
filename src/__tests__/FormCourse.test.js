import React from 'react';
import {
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import { FormCourse } from '../components/FormCourse';
import * as util from '../test-util/test-util';

describe('Tests FormCourse', () => {
  it('The text "create a fun course!" be present', () => {
    render(
      <BrowserRouter>
        <div className="desktopMenu" />
        <FormCourse
          setFalseLoad={util.func}
          location={util.mockLocation}
          setShowMenu={util.func}
          url={util.mockUrl}
          setUserErr={util.func}
          setTrueLoad={util.func}
          userId={1}
          password="password"
          id={util.mockId}
          token={util.mockToken}
          username="Javier"
          urlApi={util.mockUrlApi}
          objThunk={util.func}
          userPayload={util.func}
          tokenPayload={util.func}
          dispatchCourse={util.func}
        />
      </BrowserRouter>,
    );
    expect(screen.getByText('create a fun course!')).toBeInTheDocument();
  });
  it('Triggers dispatchCourse when user clicks in Submit', () => {
    const myFunc = jest.fn();
    render(
      <BrowserRouter>
        <div className="desktopMenu" />
        <FormCourse
          setFalseLoad={util.func}
          location={util.mockLocation}
          setShowMenu={util.func}
          url={util.mockUrl}
          setUserErr={util.func}
          setTrueLoad={util.func}
          userId={1}
          password="password"
          id={util.mockId}
          token={util.mockToken}
          username="Javier"
          urlApi={util.mockUrlApi}
          objThunk={util.func}
          userPayload={util.func}
          tokenPayload={util.func}
          dispatchCourse={myFunc}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(myFunc.mock.calls.length).toBe(1);
  });
  it('Triggers dispatchCourse when user clicks in Submit', () => {
    const myFunc = jest.fn();
    render(
      <BrowserRouter>
        <div className="desktopMenu" />
        <FormCourse
          setFalseLoad={util.func}
          location={util.mockLocation}
          setShowMenu={util.func}
          url={util.mockUrl}
          setUserErr={util.func}
          setTrueLoad={util.func}
          userId={1}
          password="password"
          id={util.mockId}
          token={util.mockToken}
          username="Javier"
          urlApi={util.mockUrlApi}
          objThunk={util.func}
          userPayload={util.func}
          tokenPayload={util.func}
          dispatchCourse={myFunc}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));
    expect(myFunc.mock.calls.length).toBe(1);
  });
  it('Makes sure render correctly', () => {
    const tree = renderer.create(
      <BrowserRouter>
        <div className="desktopMenu" />
        <FormCourse
          setFalseLoad={util.func}
          location={util.mockLocation}
          setShowMenu={util.func}
          url={util.mockUrl}
          setUserErr={util.func}
          setTrueLoad={util.func}
          userId={1}
          password="password"
          id={util.mockId}
          token={util.mockToken}
          username="Javier"
          urlApi={util.mockUrlApi}
          objThunk={util.func}
          userPayload={util.func}
          tokenPayload={util.func}
          dispatchCourse={util.func}
        />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
