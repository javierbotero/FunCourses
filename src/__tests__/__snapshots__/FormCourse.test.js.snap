// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests FormCourse Makes sure render correctly 1`] = `
Array [
  <div
    className="desktopMenu"
  />,
  <div
    className="container positionImage"
    style={
      Object {
        "backgroundImage": "url(fun1.jpeg)",
      }
    }
  >
    <nav
      className="navBar navMenu show"
    >
      <div>
        <a
          href="url"
          onClick={[Function]}
        >
          <i
            className="fa chevron-left"
          />
        </a>
      </div>
      <div>
        Create
      </div>
      <button
        className="removeButtonStyles"
        onClick={[Function]}
        type="button"
      >
        <i
          className="fa ellipsis-v"
        />
      </button>
    </nav>
    <div
      className="layer bgImage"
      id="layer"
    />
    <div
      className="form"
    >
      <header
        className="header"
      >
        <div>
          <h3>
            Javier
          </h3>
          <h4>
            create a fun course!
          </h4>
        </div>
      </header>
      <form
        className="form correctForm"
        onSubmit={[Function]}
      >
        <div>
          <label
            htmlFor="title"
          >
            <div>
              Title
            </div>
            <input
              id="title"
              onChange={[Function]}
              placeholder="Title"
              type="text"
              value=""
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="content"
          >
            <div>
              Content
            </div>
            <textarea
              className="long-text"
              id="content"
              onChange={[Function]}
              placeholder="Be concise"
              type="text"
              value=""
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="dateStart"
          >
            <div>
              Start
            </div>
            <input
              id="dateStart"
              max="4/1/2021"
              onChange={[Function]}
              type="date"
              value="4/1/2021"
            />
          </label>
          <label
            htmlFor="dateEnd"
          >
            <div>
              End
            </div>
            <input
              id="dateEnd"
              min="4/1/2021"
              onChange={[Function]}
              type="date"
              value="4/1/2021"
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="price"
          >
            <div>
              Price
            </div>
            <input
              id="price"
              onChange={[Function]}
              placeholder="Price"
              type="text"
              value=""
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="price"
          >
            <div>
              Videocall provider
            </div>
            <select
              onChange={[Function]}
              value="zoom"
            >
              <option
                value="zoom"
              >
                Zoom
              </option>
              <option
                value="meet"
              >
                Meet
              </option>
              <option
                value="livestorm"
              >
                Livestorm
              </option>
              <option
                value="zoho-meeting"
              >
                Zoho Meeting
              </option>
              <option
                value="pexip"
              >
                Pexip
              </option>
              <option
                value="poly"
              >
                Poly
              </option>
              <option
                value="Blue-jeans"
              >
                Blue Jeans
              </option>
            </select>
          </label>
        </div>
        <div>
          <label
            htmlFor="link"
          >
            <div>
              link
            </div>
            <input
              id="link"
              onChange={[Function]}
              placeholder="Paste the link"
              type="text"
              value=""
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="link"
          >
            <div>
              Status
            </div>
            <select
              onChange={[Function]}
              value="Closed"
            >
              <option
                value="Open"
              >
                Open
              </option>
              <option
                value="Closed"
              >
                Closed
              </option>
            </select>
          </label>
        </div>
        <div>
          <label
            htmlFor="main-picture"
          >
            <div>
              Main picture
            </div>
            <input
              accept="image/*"
              id="main-picture"
              name="main-picture"
              onChange={[Function]}
              type="file"
            />
          </label>
        </div>
        <div>
          <label
            htmlFor="main-picture"
          >
            <div>
              Add 3 more engaging pictures
            </div>
            <input
              accept="image/*"
              id="main-picture"
              multiple={true}
              name="main-picture"
              onChange={[Function]}
              type="file"
            />
          </label>
        </div>
        <div
          className="submit"
        >
          <input
            className="button"
            type="submit"
            value="Submit"
          />
        </div>
      </form>
    </div>
  </div>,
]
`;
